File: Data Structures/Linked Lists/Doubly Linked Lists/double-linked-list.py
# Double Linked List


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None


class DoubleLinkedList:
    def __init__(self):
        self.head = None

    def push(self, new_data):
        new_node = Node(new_data)
        new_node.next = self.head
        if self.head is not None:
            self.head.prev = new_node

        self.head = new_node

    def insertAfter(self, prev_node, data):
        if prev_node is None:
            print('Not Possible')
            return

        new_node = Node(data)
        new_node.next = prev_node.next
        prev_node.next = new_node
        new_node.prev = prev_node
        if new_node.next is not None:
            new_node.next.prev = new_node

    def printList(self, node):
        while(node is not None):
            print "%d" %(node.data),
            last = node
            node = node.next


if __name__ == '__main__':
    dllist = DoubleLinkedList()
    dllist.push(1)
    dllist.push(12)
    dllist.push(114)
    dllist.insertAfter(dllist.head.next, 8)
    print(dllist.printList(dllist.head))

File: Data Structures/Linked Lists/Doubly Linked Lists/double-linked-list.rb
class Node
    attr_accessor :val, :prev, :next

    def initialize(val, prev_node, next_node)
        @val = val
        @prev = prev_node
        @next = next_node
    end
end

class DoubleLinkedList
    def initialize(val)
        @head = Node.new(val,nil,nil)
    end

    def push(val)
        new_node = Node.new(val, nil, @head)
        if @head != nil
            @head.prev = new_node
        end
        @head = new_node
    end

    def insert_node_at_end(val)
        current_node = @head

        while current_node.next != nil
          current_node = current_node.next
        end

        new_node = Node.new(val, nil,nil)

        current_node.next = new_node
    end

    def print_list
        current = @head
        while current != nil
            puts "#{current.val}"
            current = current.next
        end
    end
end

l = DoubleLinkedList.new(3)
l.push(2)
l.insert_node_at_end(4)
l.insert_node_at_end(5)
l.insert_node_at_end(6)
l.print_list()
